src/nvim/move.c

```diff
@@ -1886,101 +1869,101 @@ void set_empty_rows(win_T *wp, int used)
 /// When scrolling scroll at least "min_scroll" lines.
 /// If "set_topbot" is true, set topline and botline first (for "zb").
 /// This is messy stuff!!!
-void scroll_cursor_bot(int min_scroll, bool set_topbot)
+void scroll_cursor_bot(win_T *wp, int min_scroll, bool set_topbot)
 {
   lineoff_T loff;
-  linenr_T old_topline = curwin->w_topline;
-  int old_skipcol = curwin->w_skipcol;
-  int old_topfill = curwin->w_topfill;
-  linenr_T old_botline = curwin->w_botline;
-  int old_valid = curwin->w_valid;
-  int old_empty_rows = curwin->w_empty_rows;
-  linenr_T cln = curwin->w_cursor.lnum;  // Cursor Line Number
-  bool do_sms = curwin->w_p_wrap && curwin->w_p_sms;
+  linenr_T old_topline = wp->w_topline;
+  int old_skipcol = wp->w_skipcol;
+  int old_topfill = wp->w_topfill;
+  linenr_T old_botline = wp->w_botline;
+  int old_valid = wp->w_valid;
+  int old_empty_rows = wp->w_empty_rows;
+  linenr_T cln = wp->w_cursor.lnum;  // Cursor Line Number
+  bool do_sms = wp->w_p_wrap && wp->w_p_sms;
 
   if (set_topbot) {
     bool set_skipcol = false;
 
     int used = 0;
-    curwin->w_botline = cln + 1;
+    wp->w_botline = cln + 1;
     loff.fill = 0;
-    for (curwin->w_topline = curwin->w_botline;
-         curwin->w_topline > 1;
-         curwin->w_topline = loff.lnum) {
-      loff.lnum = curwin->w_topline;
-      topline_back_winheight(curwin, &loff, false);
+    for (wp->w_topline = wp->w_botline;
+         wp->w_topline > 1;
+         wp->w_topline = loff.lnum) {
+      loff.lnum = wp->w_topline;
+      topline_back_winheight(wp, &loff, false);
```

Comment:
The `hasFolding()` in `topline_back_winheight()` should be changed to `hasFoldingWin()`.